### DISCUSSION TITLE
PtyMux Protocol

### DISCUSSION THREAD

**Discussion by wez** [wezterm/wezterm#4889] [Ideas]
I've been thinking about this for a couple of years now, and I'd like to get some folks thinking and talking about it.
The goal of this post is to solicit interest, and then I'll follow up by making a PR with a more detailed spec-like doc describing the protocol and we can drill into it there.

# Background

There are a several pieces of software that enable command or filesystem access to a "remote" system (where remote can actually be local, or in a container, and not just on a separate physical machine). For example, `ssh` provides a basic but solid foundation, and terminal multiplexers such as `tmux` and `zellij` provide session resumption and workspace management. Tools such as [distant](https://github.com/chipsenkbeil/distant.nvim) extend a local editor to provide convenient access to the remote system. Graphical terminal emulators (`iTerm2`, `wezterm`, [Wave](https://www.waveterm.dev/)) have or would like to have some support for remote multiplexers, either via `tmux`'s command/control protocol or via a proprietary interface.

I think there is scope for a protocol that would allow these different pieces of software to interoperate more easily, making the experience of their respective users better by enabling them to mix and match the best parts of each of those tools in their workflow.

# Quick Technical Summary

The PtyMux protocol will have at least the following:

* Process execution with the ability to stream stdin/stdout/stderr streams, send signals, obtain exit status
* PTYs can be created and processes spawned into them. Properties of the PTY, such as termios, can be reported/manipulated.
* Session/workspace layout: the ability to group/arrange PTYs into Windows, Tabs, Panes
* Server-side terminal emulation to better facilitate session resumption and functions like navigating and searching scrollback
* Filesystem access and ability to transfer files/folders in either direction. This is helpful both for transferring raw data, and also for functionality like enhanced line editing and filename completion in a graphic terminal emulator.

I plan for this to be a thing that you could talk to via eg: `ssh machine /path/to/ptymux --server` using stdin/stdout. In terms of framing the protocol, I was thinking of targeting something like [BSON](https://en.wikipedia.org/wiki/BSON) because it is inherently framed and has an efficient way to represent binary data, and is relatively easy to target from a variety of languages without requiring specialized IDL/RPC tooling.

An important goal is to not require additional open ports or bespoke authentication schemes for this; running it on top of a basic ssh connection will give maximal portability/compatibility. And for local access, running it directly on top of a unix domain socket is highly desirable.

# Where does this fit with other software?

* I think it would be great if `zellij` and/or `tmux` could either implement this protocol to expose their state, or if there could be an adapter to eg: convert `tmux -CC` into it.
* I'd like to shift wezterm's current mux protocol over to an implementation of this protocol, so its server would expose its state this way, and its client could then connect to any compliant server implementation.
* If `iTerm2` implemented support for this, it could then also connect to any compliant server implementation, and if it implemented a server, it could expose its state to other clients; imagine if you could tmux/zellij into your already running gui that has windows and tabs when you are somewhere remote and have only a basic tablet and ssh client.
* From chatting with @chipsenkbeil, I think distant would love to replace a chunk of its infrastructure with a client that speaks this protocol.

# So now what?

I have a high level picture in my mind of what I think the various request/response/pdus look like for this protocol, but haven't committed it to "paper" yet. If you've had thoughts about something similar, or have insights into things that you missed while building your own functionality, I would to hear those and see if those can be accommodated in the design.

If you're interested in reviewing and/or providing feedback on this, please comment below!

Tagging a few people that I think may be interested in this, or that may have opinions or other insights. If you're not interested in this then I apologize for the ping and you are of course welcome to unsubscribe and ignore further discussion. If you can think of others who may have interest, then please share this discussion with them.

* @nicm
* @imsnif
* @gnachman
* @sawka
* @chipsenkbeil

**Comment by imsnif** [comment]
Hey, this sounds cool. Some initial thoughts from my side about integrating this with Zellij:

1. Zellij has optional pane frames which we indicate using "content offsets" for panes. (eg. a pane with a 4 sided frame will have a content offset of 1 on all sides) - I guess these will have to be accounted for in the protocol.
2. Some Zellij panes do not run a pty but rather a plugin (so in the end, a wasm runtime). These behave a little differently (eg. they do not have an exit status and their state is cleared on each render). These are first class citizens in our multiplexer environment and are also used for things like UIs.
3. Zellij also supports floating panes, so z-index should also be accounted for.
4. Zellij does not keep tiled (non-floating) panes in an internal tree representation like most multiplexers do. We do this to allow greater freedom when resizing. I'd recommend going this way, but this is not a must on our part since we know how to translate our state to a tree structure (when possible).
5. Zellij supports suppressed panes (panes that exist and are running but do not appear on screen at the moment).
6. Zellij supports "stacked" panes, these are tiled (non-floating) panes that appear in a stack - meaning that all but one of them are just one line containing the pane title, with one pane being "expanded". The user can page through them, moving the expanded pane up or down the stack. I think these will have to be part of the protocol as well since I don't think they can be easily ignored when serializing a tiled pane state.

These are the major integration challenges that I can think of right now, I might have more when I understand the protocol better and I'd be happy to review it further and provide feedback if you'd like.

Good luck! Looking forward to reading more.

**Reply by wez** [reply]
Thanks for engaging!

re: pane layout, that makes sense. What I had in mind was allowing the client to request arbitrary layout/positioning of the panes, with the server being able to nudge that into whatever constraints it requires. Honestly, I was leaning towards the server being layout agnostic in the initial draft of this, so that the client could set the layout to anything it wanted, and then hashing out a separate "apply layout" request that would request the server to apply some named layout; this needs a bit more refinement because this is an area that doesn't have corresponding implementation in wezterm today, so I've not thought too deeply about it so far.

re: tree structure: I wasn't going to require this. I think its constraints are frustrating and limit other more creative layouts.

I hadn't considered z-index, but it's not difficult to account for that property in the protocol. 

re: non-pty panes, that's fine; in wezterm there is a concept of *domain* that I need to expose to this protocol, perhaps re-labeled as *namespace*.  In wezterm, a domain is conceptually a separate execution domain/namespace (eg: remote ssh host, remote mux, docker container, wsl instance).  wezterm also has a number of *overlays* that are essentially embedded termwiz mini-applications that are in a similar category as zellij plugins.  So, the protocol will accommodate these as well, by setting the namespace appropriately when launching them.

**Reply by gnachman** [reply]
iTerm2 has a new ssh integration feature that starts with the premise of a framed protocol over ssh, but the channel is primarily used for metadata and file transfer rather than mucking about with PTYs because I think most users would prefer to stick with their shells for launching processes. My main question about your idea is how users will launch programs? Will there be a shell? Would they need to use a wrapper to break a job into its own window (e.g., a user must run `ptymux emacs` instead of `emacs`)? Will every child job get its own PTY? Are there enough PTY devices by default on popular platforms for that to work (I believe this is a kernel parameter in Linux)? Would it make more sense for this to be a feature of tmux?

**Reply by wez** [reply]
@gnachman think of this as an alternative to the `tmux -CC` protocol, rather than an alternative to tmux itself. tmux could choose to implement this protocol, and other multiplexer clients and servers could also implement it to enable cross compatibility.

**Reply by gnachman** [reply]
I see. I think you need to tackle a bigger problem to justify making such a big change to tmux or to justify creating a new terminal multiplexer. I would be surprised if the tmux maintainers would be interested in a new protocol for its own sake. I wish tmux were more open to features like file transfer, so maybe there's something you can do with that.

There is a lot of black magic involved in making `tmux -CC` work which is probably why other terminals have not integrated with it. Unfortunately it is a fundamentally difficult problem. Any difficulty posed by the protocol is mostly due to inadequate documentation rather than technical deficits. If I were designing the system from scratch I would want to improve these areas:

* Security - there have been several CVEs because the terminal can be tricked into sending attacker-controlled text. Since the tmux protocol uses in-band signaling, an attacker-controlled host you ssh into can take advantage of that to cause your terminal to send harmful text which is handled by your own shell after closing the connection unexpectedly.
* Robustness - it would be nice to have a checksum and retry mechanism. The use of `wall` throws tmux's protocol into chaos.
* Model - The model tmux uses is of a TTY-focused UI. For example, it does not know about scrollbars and it assumes dividers between split panes are one character wide. Translating between the platform-specific UI constraints and its model is the source of many of the hardest problems. I'm not quite sure how you'd fix this while supporting multiple clients concurrently since different clients might have different scrollbar widths, etc.

For a good example of why this is hard, see the conversation about [flow control](https://github.com/tmux/tmux/issues/2217).

**Reply by wez** [reply]
FWIW, wezterm has an existing terminal multiplexer with its own remote mux protocol. I'm not advocating to build another other multiplexer; this effort is aimed at enabling/facilitating interop between existing multiplexer clients and servers.

There have been a couple of attempts to integrate `tmux -CC` into the client side of wezterm, and neither have been completed because it is difficult to integrate; as you mentioned, there is little information about it, and lots of implicit knowledge is encoded into both tmux and iterm2 which makes it hard to understand the protocol without first understanding how both of those projects work.

wezterm's own protocol is also not easy to consume (effectively impossible) outside of wezterm, so one of my primary goals around PtyMux is to shift that to a well documented and (hopefully!) easy to use protocol.

I don't require tmux to implement this proposed protocol for the effort spent on it to be worthwhile and I won't be pushing for it do so. I think it would be nice if it did, but I don't need to lobby for it to happen. I think it will be possible for someone to write an adapter between `tmux -CC` and this protocol in a separate executable if they have such a desire and are motivated.

What I think might be of interest to iTerm2 here is the potential for it to connect to remote zellij and wezterm multiplexers in addition to remote tmux multiplexers, assuming that the zellij folks decide to implement this protocol as well.  While I think it would be nice for iterm2 and zellij to ultimately adopt this protocol, those are not requirements from my perspective.

Ultimately: if no other projects choose to implement this protocol, the discussion should result in a more robust design overall, and that will be a positive outcome in my eyes. (Assuming that that discussion doesn't devolve into a brawl!)

**Reply by nicm** [reply]
The documentation for control mode hasn't improved because hardly anyone has ever asked about it. It's difficult mostly because everyone's terminal works in a different way.

I wouldn't be against adding to or replacing it if something better is available and widely adopted.

I'll wait to see what it looks like before I make comments.

It's going to be a big protocol, especially if you do the terminal emulation on the server. Even if you do it client side you will need to specify how the escape sequences are to be interpreted, something that has proved difficult up to now.

As George mentioned, don't forget about flow control!

**Reply by gnachman** [reply]
I'll add the other thing that causes headaches is client and server having different versions. Think carefully about forward and backward compatibility. In iTerm2<->tmux the client has to know about more or less every version of tmux that has ever existed to take advantage of newly added features, polyfill those features for older versions (if possible), and to work around bugs. I think that portability is going to be painful if there are multiple server implementations with different feature sets or bugs.

**Comment by scabala** [comment]
Hi @wez,
this idea for protocol looks very cool! It would be a great booster for custom terminals and terminal multiplexers implementation if something like that existed. Did you manage to work on it?

**Reply by wez** [reply]
I have some designs and a partial implementation inside wezterm. There doesn't seem to be sufficient interest from other terminal authors to collaborate, so that aspect has stalled.

**Reply by scabala** [reply]
That's what I suspected from this thread. Bummer. Hopefully it will be picked up some day.

**Comment by cyclic-elevator** [comment]
@wez any further development on this?

Is it safe to assume that given the current state of tmux integration in wezterm, there is little motivation from the wezterm user community to push for a better protocol?

I am curious if folks have compiled a list of wish list based on experience with iterm2/wezterm+tmux so far.

In addition to wezterm, such a protocol can lower the integration cost of zellij/tmux in vscode/zed/alacritty/xterm.js etc.

**Reply by wez** [reply]
Nothing more to add since https://github.com/wezterm/wezterm/discussions/4889#discussioncomment-12201633
I would like to continue to evolve this inside wezterm, but I have not had much time to spend on it.

**Reply by cyclic-elevator** [reply]
Can we do some AI-assisted spec/proposal iterations/explorations? Something along the lines of:
- extract current tmux integration specs out of iterm2 and tmux repos
- synthesize new PtyMux spec using extracted tmux spec and proposed changes
- explore scope/impact/viability using the synthesized PtyMux spec as input and target tmux/zellij repos for proposed changes

**Reply by wez** [reply]
No thanks, I don't want that.

**Comment by scabala** [comment]
I was wondering on one point:
> Server-side terminal emulation to better facilitate session resumption and functions like navigating and searching scrollback

Is it really necessary? I recently saw [shpool](https://github.com/shell-pool/shpool) - which is not perfect or being advertised by me - and from what I understand, it does not do terminal emulation - or at least very minimal portion of it. Like `dtach` or `abduco`, stdout and stderr from process is streamed as-is to the client without any interpretation. From what I saw around in terminal world, emulating terminal and parsing all the escape sequences is _very_, _very_ complex topic and everybody does it a little bit different. 

Obviously, then each client's responsibility would be to emulate terminal but it allow greater flexibility for terminal emulator and multiplexer authors.

**Reply by wez** [reply]
It's not necessary; it's an optional feature in my protocol design. But keep in mind that the most likely thing to implement the full spec is a terminal emulator that has multiplexing support, so the difficult parts of emulation would be largely a solved problem already.

**Comment by gtrak** [comment]
This reminds me a bit of clojure/nREPL, which is a control protocol between clients (like emacs) and REPL servers that also can stream response data.  Maybe you could get some inspiration there? https://nrepl.org/nrepl/design/overview.html

Specifically, it stands out to me that if something here doesn't implement the full spec, it needs a way to advertise what it does implement, or the spec needs to have an extension mechanism like nREPL does. Extra functionality like completions was implemented as middleware.

I came across this thread by looking for a way to drive zellij more programmatically.  I like the UX and I want to integrate it into other tools to manage long-running background tasks like running tests or project docker containers.
